src/pubsub.ts

```diff
@@ -541,32 +546,125 @@ export class PubSub {
+    callback = typeof optsOrCallback === 'function' ? optsOrCallback : callback;
+
+    this.request<google.pubsub.v1.IDetachSubscriptionRequest>(
+      {
+        client: 'PublisherClient',
+        method: 'detachSubscription',
+        reqOpts,
+        gaxOpts: gaxOpts as CallOptions,
+      },
+      callback!
+    );
+  }
+
   /**
    * Determine the appropriate endpoint to use for API requests, first trying
-   * the local `apiEndpoint` parameter. If the `apiEndpoint` parameter is null
-   * we try Pub/Sub emulator environment variable (PUBSUB_EMULATOR_HOST),
-   * otherwise the default JSON API.
+   * the local `apiEndpoint` parameter. If the `apiEndpoint` parameter is null,
+   * we try the standard `gcloud alpha pubsub` environment variable
+   * (CLOUDSDK_API_ENDPOINT_OVERRIDES_PUBSUB), and failing that, we try the
+   * Pub/Sub emulator environment variable (PUBSUB_EMULATOR_HOST).
+   * Otherwise the default production API is used.
    *
    * @private
    */
   determineBaseUrl_() {
-    const apiEndpoint = this.options.apiEndpoint;
-    if (!apiEndpoint && !process.env.PUBSUB_EMULATOR_HOST) {
+    // We allow an override from the client object options, or from
+    // one of these variables. The CLOUDSDK variable is provided for
+    // compatibility with the `gcloud alpha` utility.
+    const gcloudVarName = 'CLOUDSDK_API_ENDPOINT_OVERRIDES_PUBSUB';
+    const emulatorVarName = 'PUBSUB_EMULATOR_HOST';
+    const apiEndpoint =
+      this.options.apiEndpoint ||
+      process.env[gcloudVarName] ||
+      process.env[emulatorVarName];
+    if (!apiEndpoint) {
       return;
     }
 
-    const grpcInstance = this.options.grpc || gax.grpc;
-    const baseUrl = apiEndpoint || process.env.PUBSUB_EMULATOR_HOST;
-    const leadingProtocol = new RegExp('^https*://');
+    // Parse the URL into a hostname and port, if possible.
+    const leadingProtocol = new RegExp('^https?://');
     const trailingSlashes = new RegExp('/*$');
-    const baseUrlParts = baseUrl!
+    const baseUrlParts = apiEndpoint!
       .replace(leadingProtocol, '')
       .replace(trailingSlashes, '')
       .split(':');
     this.options.servicePath = baseUrlParts[0];
-    this.options.port = baseUrlParts[1];
-    this.options.sslCreds = grpcInstance.credentials.createInsecure();
-    this.isEmulator = true;
+    if (!baseUrlParts[1]) {
+      // No port was given -- figure it out from the protocol.
+      if (apiEndpoint!.startsWith('https')) {
+        this.options.port = 443;
+      } else if (apiEndpoint!.startsWith('http')) {
+        this.options.port = 80;
+      } else {
+        this.options.port = undefined;
+      }
+    } else {
+      this.options.port = parseInt(baseUrlParts[1], 10);
+    }
+
+    // If this looks like a GCP URL of some kind, don't go into emulator
+    // mode. Otherwise, supply a fake SSL provider so a real cert isn't
+    // required for running the emulator.
+    const officialUrlMatch = this.options.servicePath!.match(
```

Comment:
`this.options.servicePath!.endsWith('.googleapis.com')` is so readable for regex-impaired folk like me